include/net/xdp.h

```diff
@@ -131,6 +153,13 @@ xdp_prepare_buff(struct xdp_buff *xdp, unsigned char *hard_start,
 	xdp->data = data;
 	xdp->data_end = data + data_len;
 	xdp->data_meta = meta_valid ? data : data + 1;
+
+	if (meta_valid) {
+		/* If a driver supports metadata, it needs at least enough
+		 * headroom to store traits.
+		 */
+		BUG_ON(traits_init(xdp_traits(xdp), xdp->data_meta) != 0);
```

Comment:
Explanation: why performance issues for XDP use-case

Both `BUG_ON` and `WARN_ON` are smart and adds a `UD2` instruction
- this inst cause the CPU instruction prefetcher to avoid loading code into instruction-cache
- makes sense as we expect this code should never be run

Sounds like a good optimization, but not for this XDP use-case
- because `xdp_prepare_buff` get inlined into all XDP NIC drivers
- and we see this `UD2` inst often get placed wrongly by compiler
- this have in the past lead to hard-to-find performance regressions in NIC drivers